backends-velox/src/main/scala/org/apache/gluten/execution/ColumnarCollectLimitExec.scala

```diff
@@ -32,88 +32,94 @@ import org.apache.spark.sql.vectorized.ColumnarBatch
    * Either retain the entire batch if it fits within the remaining limit, or prune it if it
-   * partially exceeds the remaining limit.
+   * partially exceeds the remaining limit/offset.
    */
-  private def collectLimitedRows(
-      partitionIter: Iterator[ColumnarBatch],
-      limit: Int
-  ): Iterator[ColumnarBatch] = {
-    if (partitionIter.isEmpty) {
-      return Iterator.empty
-    }
-    new Iterator[ColumnarBatch] {
+  private def collectWithOffsetAndLimit(
+                                   inputIter: Iterator[ColumnarBatch],
+                                   offset: Int,
+                                   limit: Int): Iterator[ColumnarBatch] = {
+
+    val unlimited = limit < 0
+    var rowsToSkip = math.max(offset, 0)
+    var rowsToCollect = if (unlimited) Int.MaxValue else limit
 
-      private var rowsCollected = 0
+    new Iterator[ColumnarBatch] {
       private var nextBatch: Option[ColumnarBatch] = None
 
       override def hasNext: Boolean = {
-        nextBatch.isDefined || fetchNext()
+        nextBatch.isDefined || fetchNextBatch()
       }
 
       override def next(): ColumnarBatch = {
-        if (!hasNext) {
-          throw new NoSuchElementException("No more batches available.")
-        }
+        if (!hasNext) throw new NoSuchElementException("No more batches available.")
         val batch = nextBatch.get
         nextBatch = None
         batch
       }
 
       /**
-       * Attempt to fetch the next batch from the underlying iterator if we haven't yet hit the
-       * limit. Returns true if we found a new batch, false otherwise.
+       * Advance the iterator until we find a batch (possibly sliced)
+       * that we can return, or exhaust the input.
        */
-      private def fetchNext(): Boolean = {
-        if (rowsCollected >= limit || !partitionIter.hasNext) {
-          return false
-        }
-
-        val currentBatch = partitionIter.next()
-        val currentBatchRowCount = currentBatch.numRows()
-        val remaining = limit - rowsCollected
-
-        if (currentBatchRowCount <= remaining) {
-          rowsCollected += currentBatchRowCount
-          ColumnarBatches.retain(currentBatch)
-          nextBatch = Some(currentBatch)
-        } else {
-          val prunedBatch = VeloxColumnarBatches.slice(currentBatch, 0, remaining)
-          rowsCollected += remaining
-          nextBatch = Some(prunedBatch)
+      private def fetchNextBatch(): Boolean = {
+        if (rowsToCollect <= 0) return false
+
+        while (inputIter.hasNext) {
+          val batch = inputIter.next()
+          val batchSize = batch.numRows()
+
+          if (rowsToSkip >= batchSize) {
+            rowsToSkip -= batchSize
+          } else {
+            val startIndex = rowsToSkip
+            val leftoverAfterSkip = batchSize - startIndex
+            rowsToSkip = 0
+
+            val needed = math.min(rowsToCollect, leftoverAfterSkip)
```

Comment:
if needed <= remaining, we still need this logic, may return the total batch instead of sliced batch
```
f (currentBatchRowCount <= remaining) {
          rowsCollected += currentBatchRowCount
          ColumnarBatches.retain(currentBatch)
          nextBatch = Some(currentBatch)
        } else {
          val prunedBatch = VeloxColumnarBatches.slice(currentBatch, 0, remaining)
```